mlflow/pipelines/steps/train.py

```diff
@@ -125,88 +123,136 @@ def _run(self, output_directory):
         }
-        with mlflow.start_run(tags=tags) as run:
-            estimator.fit(X_train, y_train)
-
-            if hasattr(estimator, "best_score_"):
-                mlflow.log_metric("best_cv_score", estimator.best_score_)
-            if hasattr(estimator, "best_params_"):
-                mlflow.log_params(estimator.best_params_)
-
-            # TODO: log this as a pyfunc model
-            code_paths = [os.path.join(self.pipeline_root, "steps")]
-            estimator_schema = infer_signature(X_train, estimator.predict(X_train.copy()))
-            logged_estimator = mlflow.sklearn.log_model(
-                estimator,
-                f"{self.name}/estimator",
-                signature=estimator_schema,
-                code_paths=code_paths,
-            )
-            # Create a pipeline consisting of the transformer+model for test data evaluation
-            with open(transformer_path, "rb") as f:
-                transformer = cloudpickle.load(f)
-            mlflow.sklearn.log_model(transformer, "transform/transformer", code_paths=code_paths)
-            model = make_pipeline(transformer, estimator)
-            model_schema = infer_signature(raw_X_train, model.predict(raw_X_train.copy()))
-            model_info = mlflow.sklearn.log_model(
-                model, f"{self.name}/model", signature=model_schema, code_paths=code_paths
+
+        if self.step_config["tuning_enabled"]:
+            # gate all HP tuning code within this condition
+            tuning_params = self.step_config["tuning"]
+            # import hyperopt or throw error
+            try:
+                from hyperopt import hp, fmin
+            except ModuleNotFoundError:
+                raise MlflowException(
+                    "Hyperopt not installed and is required if tuning is enabled",
+                    error_code=BAD_REQUEST,
+                )
+
+            # wrap training in objective fn
+            def objective(args):  # pylint: disable=unused-argument
+                return None
+
+            # construct hp search space from yaml
+            search_space = {}
+            params = tuning_params["parameters"]
+            for param_name, param_details in params.items():
+                if "values" in param_details:
+                    search_space[param_name] = hp.choice(param_name, **param_details)
+                elif "distribution" in param_details:
+                    hp_tuning_fn = getattr(hp, param_details["distribution"])
+                    param_details_to_pass = param_details.copy()
+                    param_details_to_pass.pop("distribution")
+                    search_space[param_name] = hp_tuning_fn(param_name, **param_details_to_pass)
+                else:
+                    raise MlflowException(
+                        f"Parameter {param_name} must contain the arguments specified "
+                        f"in the hyperopt parameter expressions",
```

Comment:
Can we bit a bit more specific here and indicate that the tuning parameter configuration must contain either a "distribution" or "values" key with particular contents (explain what the contents are)?